src/main/java/org/apache/sling/scripting/jsp/PrecompiledJSPRunner.java

```diff
@@ -0,0 +1,136 @@
+import javax.servlet.ServletException;
+
+import org.apache.sling.api.SlingException;
+import org.apache.sling.api.SlingIOException;
+import org.apache.sling.api.SlingServletException;
+import org.apache.sling.api.scripting.SlingBindings;
+import org.apache.sling.commons.compiler.source.JavaEscapeHelper;
+import org.apache.sling.scripting.bundle.tracker.BundledRenderUnit;
+import org.apache.sling.scripting.jsp.jasper.compiler.JspRuntimeContext;
+import org.apache.sling.scripting.jsp.jasper.runtime.AnnotationProcessor;
+import org.apache.sling.scripting.jsp.jasper.runtime.HttpJspBase;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Component(
+        immediate = true,
+        service = {}
+        /*
+         * this component will register itself as a service only if the org.apache.sling.scripting.bundle.tracker API is present
+         */
+        )
+public class PrecompiledJSPRunner {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PrecompiledJSPRunner.class);
+
+    private final ServiceRegistration<?> serviceRegistration;
+
+    @Activate
+    public PrecompiledJSPRunner(BundleContext bundleContext) {
+        serviceRegistration = register(bundleContext);
+    }
+
+    @Deactivate
+    public void deactivate() {
+        if (serviceRegistration != null) {
+            serviceRegistration.unregister();
+        }
+    }
+
+    boolean callPrecompiledJSP(JspRuntimeContext.JspFactoryHandler jspFactoryHandler, JspServletConfig jspServletConfig,
+                               SlingBindings bindings) {
+        boolean found = false;
+        BundledRenderUnit bundledRenderUnit = (BundledRenderUnit) bindings.get(BundledRenderUnit.VARIABLE);
+        if (bundledRenderUnit != null && bundledRenderUnit.getUnit() instanceof HttpJspBase) {
+            found = true;
+            HttpJspBase jsp = (HttpJspBase) bundledRenderUnit.getUnit();
+            PrecompiledServletConfig servletConfig = new PrecompiledServletConfig(jspServletConfig, bundledRenderUnit);
+            try {
+                jspFactoryHandler.incUsage();
+                AnnotationProcessor annotationProcessor =
+                        (AnnotationProcessor) jspServletConfig.getServletContext().getAttribute(AnnotationProcessor.class.getName());
+                if (annotationProcessor != null) {
+                    annotationProcessor.processAnnotations(jsp);
+                    annotationProcessor.postConstruct(jsp);
+                }
+                if (jsp.getServletConfig() == null) {
+                    jsp.init(servletConfig);
+                }
+                jsp.service(bindings.getRequest(), bindings.getResponse());
```

Comment:
I think all of the above init should only be done once per script, for subsequent requests to the same script only jsp.service() must be called

Reply:
sorry for being painful here, with the above you create a bottleneck as it synchronizes all parallel requests, regardless of the jsp and regardless whether the jsp has already been initialized.
So you might want to do something like
```
if (jsp.getServletConfig() == null) {
  synchronized ( jsp ) {
     if (jsp.getServletConfig() == null) {
      // DO IT
    }
  }
}
```